### ISSUE TITLE
HuggingfaceFlashModel Optimization Guide

### ISSUE THREAD

**Issue by ChangWenhan** [EasyJailbreak/EasyJailbreak#48]
# HuggingfaceFlashModel Optimization Guide

## Model Loading Optimizations

1. **FlashAttention 2 Support**
   - Enabled via `attn_implementation="flash_attention_2"` parameter
   - Automatic fallback mechanism when FlashAttention fails
   - Detailed logging of loading status

2. **BetterTransformer Support**
   - Alternative optimization when FlashAttention is unavailable
   - Enabled via `model.to_bettertransformer()` for PyTorch native attention optimization

3. **Torch.compile Acceleration**
   - Available for PyTorch 2.0+
   - Uses `reduce-overhead` mode to balance compilation time and inference speed
   - Automatic CUDA availability detection

4. **Efficient Memory Management**
   - `low_cpu_mem_usage=True` reduces CPU memory consumption
   - `device_map="auto"` for automatic device allocation

## Tokenizer Optimizations

1. **Fast Tokenizer**
   - Enabled via `use_fast=True` for faster tokenizer implementation

2. **Efficient Batch Processing**
   - `padding_side='left'` for optimized batch inference
   - Automatic pad_token configuration

## Inference Optimizations

1. **Generation Configuration**
   - Default KV cache enabled (`use_cache=True`)
   - Flexible generation parameter merging

2. **Efficient Inference Mode**
   - Uses `@torch.inference_mode()` instead of `@torch.no_grad()` for better performance
   - Automatic attention_mask generation

3. **Batch Generation**
   - Supports batch prompt processing
   - Automatic padding and truncation
   - Decodes only newly generated tokens to avoid redundant decoding

**Comment by ChangWenhan** [comment]
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from fastchat.conversation import get_conv_template
import logging
from easyjailbreak.models.model_base import WhiteBoxModelBase

class HuggingfaceFlashModel(WhiteBoxModelBase):
    def __init__(self, model_name_or_path, model_name, tokenizer_name_or_path=None,
                 dtype=torch.float16, flash_attn=True, torch_compile=True, 
                 use_bettertransformer=False, use_cache=True, **generation_config):
        self.model_name = model_name
        self.generation_config = generation_config
        
        # 设置默认生成参数以提升速度
        if 'use_cache' not in self.generation_config:
            self.generation_config['use_cache'] = use_cache

        # ======= Load model =======
        model_kwargs = {
            "trust_remote_code": True,
            "low_cpu_mem_usage": True,
            "device_map": "auto",
            "torch_dtype": dtype,
        }

        # FlashAttention 2 support (通过加载参数启用)
        if flash_attn:
            try:
                model_kwargs["attn_implementation"] = "flash_attention_2"
                logging.info("[INFO] Attempting to load with FlashAttention 2")
            except Exception as e:
                logging.warning(f"[WARN] FlashAttention parameter setup failed: {e}")

        try:
            model = AutoModelForCausalLM.from_pretrained(model_name_or_path, **model_kwargs)
            logging.info("[INFO] Model loaded successfully")
        except Exception as e:
            # 如果 FlashAttention 加载失败，回退到标准加载
            if flash_attn and "attn_implementation" in model_kwargs:
                logging.warning(f"[WARN] Failed to load with FlashAttention, falling back: {e}")
                model_kwargs.pop("attn_implementation", None)
                model = AutoModelForCausalLM.from_pretrained(model_name_or_path, **model_kwargs)
            else:
                raise e

        # BetterTransformer (PyTorch native attention优化)
        if use_bettertransformer and not flash_attn:
            try:
                model = model.to_bettertransformer()
                logging.info("[INFO] BetterTransformer enabled")
            except Exception as e:
                logging.warning(f"[WARN] BetterTransformer failed: {e}")

        # 设置为eval模式
        model.eval()

        # Torch.compile (PyTorch 2.0+)
        if torch_compile and torch.__version__ >= "2" and torch.cuda.is_available():
            try:
                # 使用 max-autotune 模式可能获得更好性能，但编译时间更长
                model = torch.compile(model, mode="reduce-overhead", fullgraph=False)
                logging.info("[INFO] Torch.compile applied for acceleration")
            except Exception as e:
                logging.warning(f"[WARN] Torch.compile failed: {e}")

        # ======= Load tokenizer =======
        if tokenizer_name_or_path is None:
            tokenizer_name_or_path = model_name_or_path
        tokenizer = AutoTokenizer.from_pretrained(
            tokenizer_name_or_path, 
            trust_remote_code=True, 
            use_fast=True,  # 使用 Fast Tokenizer
            padding_side='left'  # 批量推理时左侧padding更高效
        )
        if tokenizer.pad_token is None:
            tokenizer.pad_token = tokenizer.eos_token
            tokenizer.pad_token_id = tokenizer.eos_token_id
            model.config.pad_token_id = tokenizer.pad_token_id

        # ======= Init WhiteBoxModelBase =======
        super().__init__(model, tokenizer)

        # ======= Conversation template =======
        try:
            self.conversation = get_conv_template(model_name)
        except KeyError:
            logging.warning(f"[WARN] Invalid conversation template name: {model_name}")
            self.conversation = get_conv_template("vicuna_v1.1")  # fallback

        self.format_str = self._create_format_str()

    # ======= Conversation helpers =======
    def _create_format_str(self):
        self.conversation.messages = []
        self.conversation.append_message(self.conversation.roles[0], "{prompt}")
        self.conversation.append_message(self.conversation.roles[1], "{response}")
        fmt = self.conversation.get_prompt()
        self.conversation.messages = []
        return fmt

    def _build_prompt(self, messages):
        self.conversation.messages = []
        for i, msg in enumerate(messages):
            self.conversation.append_message(self.conversation.roles[i % 2], msg)
        self.conversation.append_message(self.conversation.roles[-1], None)
        return self.conversation.get_prompt()

    # ======= Implement abstract methods from WhiteBoxModelBase =======
    @torch.inference_mode()  # 使用 inference_mode 替代 no_grad，更快
    def generate(self, messages, **kwargs):
        if isinstance(messages, str):
            messages = [messages]
        prompt = self._build_prompt(messages)
        
        # 使用更高效的tokenization
        inputs = self.tokenizer(
            prompt, 
            return_tensors="pt", 
            add_special_tokens=False,
            padding=False,
            truncation=False
        )
        input_ids = inputs.input_ids.to(self.model.device)
        
        # 合并生成配置
        gen_kwargs = {**self.generation_config, **kwargs}
        gen_kwargs["input_ids"] = input_ids
        
        # 如果没有指定attention_mask，添加它
        if "attention_mask" not in gen_kwargs:
            gen_kwargs["attention_mask"] = inputs.attention_mask.to(self.model.device)
        
        output_ids = self.model.generate(**gen_kwargs)
        return self.tokenizer.decode(output_ids[0][input_ids.shape[1]:], skip_special_tokens=True)

    @torch.inference_mode()
    def batch_generate(self, messages_list, **kwargs):
        """批量生成，提升吞吐量"""
        prompts = []
        for messages in messages_list:
            if isinstance(messages, str):
                messages = [messages]
            prompts.append(self._build_prompt(messages))
        
        # 批量tokenization with padding
        inputs = self.tokenizer(
            prompts,
            return_tensors="pt",
            padding=True,
            add_special_tokens=False,
            truncation=True,
            max_length=2048  # 根据需要调整
        )
        inputs = {k: v.to(self.model.device) for k, v in inputs.items()}
        
        # 合并生成配置
        gen_kwargs = {**self.generation_config, **kwargs}
        gen_kwargs.update(inputs)
        
        output_ids = self.model.generate(**gen_kwargs)
        
        # 只解码新生成的部分
        input_length = inputs["input_ids"].shape[1]
        outputs = []
        for output in output_ids:
            text = self.tokenizer.decode(output[input_length:], skip_special_tokens=True)
            outputs.append(text)
        
        return outputs

    def instance2str(self, instance, *args, **kwargs):
        return str(instance)

    @property
    def device(self):
        return self.model.device

    @property
    def embed_layer(self):
        for m in self.model.modules():
            if isinstance(m, torch.nn.Embedding):
                return m
        return None

    @property
    def vocab_size(self):
        return self.model.config.vocab_size

    @property
    def bos_token_id(self):
        return self.tokenizer.bos_token_id

    @property
    def eos_token_id(self):
        return self.tokenizer.eos_token_id

    @property
    def pad_token_id(self):
        return self.tokenizer.pad_token_id

    def __call__(self, *args, **kwargs):
        return self.model(*args, **kwargs)

    def batch_encode(self, *args, **kwargs):
        return self.tokenizer(*args, **kwargs)

    def batch_decode(self, *args, **kwargs):
        return self.tokenizer.batch_decode(*args, **kwargs)


def from_pretrained_flash(model_name_or_path, model_name, tokenizer_name_or_path=None,
                          dtype=torch.float16, flash_attn=True, torch_compile=True,
                          use_bettertransformer=False, use_cache=True, **generation_config):
    """
    快速加载模型的工厂函数
    
    Args:
        model_name_or_path: 模型路径或HF模型ID
        model_name: 对话模板名称
        tokenizer_name_or_path: tokenizer路径（默认与model相同）
        dtype: 模型精度 (float16, bfloat16, float32)
        flash_attn: 是否使用FlashAttention 2
        torch_compile: 是否使用torch.compile加速
        use_bettertransformer: 是否使用BetterTransformer (不与flash_attn同时使用)
        use_cache: 是否使用KV cache (推荐True)
        **generation_config: 生成参数
    """
    return HuggingfaceFlashModel(
        model_name_or_path=model_name_or_path,
        model_name=model_name,
        tokenizer_name_or_path=tokenizer_name_or_path,
        dtype=dtype,
        flash_attn=flash_attn,
        torch_compile=torch_compile,
        use_bettertransformer=use_bettertransformer,
        use_cache=use_cache,
        **generation_config
    )
```

**Comment by ChangWenhan** [comment]
I have made some modifications to this model loading class to make the model's inference speed faster compared to the original source code (`easyjailbreak.models.huggingface_model.from_pretrained`).

It's actually possible to further optimize the model through 8-bit quantization or loading it via vLLM, which I may try to implement in the future.

Additionally, while conducting PAIR experiments, I encountered some issues. For instance, I wanted to run experiments through both locally deployed APIs and ByteDance's APIs, but faced template mismatch problems with `chat_template`. This doesn't seem very convenient to modify.

**Comment by BeyonderXX** [comment]
> I have made some modifications to this model loading class to make the model's inference speed faster compared to the original source code (`easyjailbreak.models.huggingface_model.from_pretrained`).
> 
> It's actually possible to further optimize the model through 8-bit quantization or loading it via vLLM, which I may try to implement in the future.
> 
> Additionally, while conducting PAIR experiments, I encountered some issues. For instance, I wanted to run experiments through both locally deployed APIs and ByteDance's APIs, but faced template mismatch problems with `chat_template`. This doesn't seem very convenient to modify.

Thanks for opening this issue and sharing your thoughts on improving the Huggingface FlashModel optimization guide!
This sounds like a valuable addition to the project — performance optimization and best practices would definitely help other users.

If you’re interested, feel free to open a **Pull Request** to contribute a draft of this guide (either as a new Markdown doc under `docs/` or as an example script).

You can start with a simple outline covering:

* Key optimization methods (e.g. FlashAttention, BetterTransformer, quantization)
* Example usage in EasyJailbreak
* Performance comparison or benchmark results (optional)

I’ll be happy to review and help refine the content.

Thanks again for your contribution and initiative!


Xiao Wang
